File: go/io-benchmark.go
package main

import (
	"fmt"
	"math/rand"
	"os"
	"sync"
	"time"
)

func createTestFiles(count int, taskID string) error {
	var wg sync.WaitGroup
	errChan := make(chan error, count)
	
	for i := 0; i < count; i++ {
		wg.Add(1)
		go func(i int) {
			defer wg.Done()
			filename := fmt.Sprintf("temp_%s_%d.txt", taskID, i)
			content := fmt.Sprintf("Test file %d content: %f", i, rand.Float64())
			if err := os.WriteFile(filename, []byte(content), 0644); err != nil {
				errChan <- err
			}
		}(i)
	}
	
	wg.Wait()
	close(errChan)
	
	if len(errChan) > 0 {
		return <-errChan
	}
	return nil
}

func readTestFiles(count int, taskID string) ([]string, error) {
	var wg sync.WaitGroup
	results := make([]string, count)
	errChan := make(chan error, count)
	
	for i := 0; i < count; i++ {
		wg.Add(1)
		go func(i int) {
			defer wg.Done()
			filename := fmt.Sprintf("temp_%s_%d.txt", taskID, i)
			
			// リトライロジック
			var content []byte
			var err error
			for retry := 0; retry < 10; retry++ {
				content, err = os.ReadFile(filename)
				if err == nil {
					results[i] = string(content)
					return
				}
				if retry < 9 {
					time.Sleep(20 * time.Millisecond)
				}
			}
			errChan <- err
		}(i)
	}
	
	wg.Wait()
	close(errChan)
	
	if len(errChan) > 0 {
		return nil, <-errChan
	}
	return results, nil
}

func cleanupTestFiles(count int, taskID string) {
	var wg sync.WaitGroup
	for i := 0; i < count; i++ {
		wg.Add(1)
		go func(i int) {
			defer wg.Done()
			filename := fmt.Sprintf("temp_%s_%d.txt", taskID, i)
			os.Remove(filename) // Ignore errors
		}(i)
	}
	wg.Wait()
}

func simulateNetworkDelay(ms int) {
	time.Sleep(time.Duration(ms) * time.Millisecond)
}

func ioIntensiveTask(taskID string) (int, error) {
	fileCount := 50
	networkCalls := 20
	
	// Create files concurrently
	if err := createTestFiles(fileCount, taskID); err != nil {
		return 0, err
	}
	
	// Short wait to ensure filesystem writes complete
	time.Sleep(5 * time.Millisecond)
	
	// Simulate network calls with delays
	var wg sync.WaitGroup
	for i := 0; i < networkCalls; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			delay := 10 + rand.Intn(20)
			simulateNetworkDelay(delay)
		}()
	}
	
	// Read files concurrently while network calls are happening
	fileContents, err := readTestFiles(fileCount, taskID)
	wg.Wait() // Wait for network calls to complete
	
	// Cleanup
	cleanupTestFiles(fileCount, taskID)
	
	if err != nil {
		return 0, err
	}
	
	return len(fileContents), nil
}

func RunIOBenchmark(taskCount int) {
	fmt.Println("Go I/O benchmark start")

	start := time.Now()

	var wg sync.WaitGroup
	for i := 0; i < taskCount; i++ {
		wg.Add(1)
		go func(i int) {
			defer wg.Done()
			taskID := fmt.Sprintf("task%d", i)
			_, err := ioIntensiveTask(taskID)
			if err != nil {
				fmt.Printf("Task %d failed: %v\n", i+1, err)
			} else {
				fmt.Printf("Task %d done\n", i+1)
			}
		}(i)
	}

	wg.Wait()
	elapsed := time.Since(start)
	fmt.Printf("All %d tasks done in %v\n", taskCount, elapsed)
}

File: rust/src/io_benchmark.rs
use std::time::{Duration, Instant};
use tokio::time::sleep;

async fn create_test_files(count: usize, task_id: &str) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    let mut handles = Vec::new();
    let task_id = task_id.to_string();
    
    for i in 0..count {
        let task_id = task_id.clone();
        let handle = tokio::spawn(async move {
            let content = format!("Test file {} content with some data: {}", i, rand::random::<f64>());
            let filename = format!("temp_{}_{}.txt", task_id, i);
            tokio::fs::write(filename, content).await
        });
        handles.push(handle);
    }
    
    for handle in handles {
        handle.await??;
    }
    
    Ok(())
}

async fn read_test_files(count: usize, task_id: &str) -> Result<Vec<String>, Box<dyn std::error::Error + Send + Sync>> {
    let mut handles = Vec::new();
    let task_id = task_id.to_string();
    
    for i in 0..count {
        let task_id = task_id.clone();
        let handle = tokio::spawn(async move {
            let filename = format!("temp_{}_{}.txt", task_id, i);
            
            // リトライロジック
            for retry in 0..10 {
                match tokio::fs::read_to_string(&filename).await {
                    Ok(content) => return Ok(content),
                    Err(_e) if retry < 9 => {
                        sleep(Duration::from_millis(10)).await;
                    }
                    Err(e) => return Err(e),
                }
            }
            unreachable!()
        });
        handles.push(handle);
    }
    
    let mut results = Vec::new();
    for handle in handles {
        results.push(handle.await??);
    }
    
    Ok(results)
}

async fn cleanup_test_files(count: usize, task_id: &str) {
    let mut handles = Vec::new();
    let task_id = task_id.to_string();
    
    for i in 0..count {
        let task_id = task_id.clone();
        let handle = tokio::spawn(async move {
            let filename = format!("temp_{}_{}.txt", task_id, i);
            let _ = tokio::fs::remove_file(filename).await; // Ignore errors
        });
        handles.push(handle);
    }
    
    for handle in handles {
        let _ = handle.await;
    }
}

async fn simulate_network_delay(ms: u64) {
    sleep(Duration::from_millis(ms)).await;
}

async fn io_intensive_task(task_id: &str) -> Result<usize, Box<dyn std::error::Error + Send + Sync>> {
    let file_count = 50;
    let network_calls = 20;
    
    // Create files concurrently
    create_test_files(file_count, task_id).await?;
    
    // Simulate network calls with delays
    let mut network_handles = Vec::new();
    for _ in 0..network_calls {
        let handle = tokio::spawn(async {
            let delay = 10 + (rand::random::<f64>() * 20.0) as u64;
            simulate_network_delay(delay).await;
        });
        network_handles.push(handle);
    }
    
    // Read files concurrently while network calls are happening
    let file_contents_future = read_test_files(file_count, task_id);
    let network_future = async {
        for handle in network_handles {
            let _ = handle.await;
        }
    };
    
    let (file_contents, _) = tokio::join!(file_contents_future, network_future);
    let file_contents = file_contents?;
    
    // Cleanup
    cleanup_test_files(file_count, task_id).await;
    
    Ok(file_contents.len())
}

pub async fn run_io_benchmark(task_count: usize) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    println!("Rust I/O benchmark start");
    
    let start = Instant::now();
    
    let mut handles = Vec::new();
    for i in 0..task_count {
        let handle = tokio::spawn(async move {
            let task_id = format!("task{}", i);
            match io_intensive_task(&task_id).await {
                Ok(_) => println!("Task {} done", i + 1),
                Err(e) => println!("Task {} failed: {}", i + 1, e),
            }
        });
        handles.push(handle);
    }
    
    for handle in handles {
        let _ = handle.await;
    }
    
    let duration = start.elapsed();
    println!("All {} tasks done in {:?}", task_count, duration);
    
    Ok(())
}